### PR TITLE
Fix overloading for torch.nn and Functional modules

### PR DESCRIPTION
**Pull request by Syzygianinfern0**
This PR fixes the following
- By #3253 the automatic overloading of torch modules was lost.
- By setting up the decorators and an efficient \_\_init\_\_.py, the functionality is restored.
- Also includes the overloading of AvgPooling and Conv2d from torch.nn which were earlier missing
- Move tests from test_precision related to tor APIs to test_nn.py and test_functional.py

### CODE REVIEW

File: syft/frameworks/torch/tensors/interpreters/precision.py [hunk 1]
```diff
@@ -754,27 +755,8 @@ def dot(self, other):
...
-                    """
-                    return torch.nn.functional.native_linear(*args)
-
-                module.linear = linear
-
-            module.functional = functional
-
         # Modules should be registered just like functions
-        module.nn = nn
+        module.nn = nn  # Handles all the overloading properly
```
Comment (LaRiffle) [inline syft/frameworks/torch/tensors/interpreters/precision.py]: I love that

Comment (iamtrask) [reply syft/frameworks/torch/tensors/interpreters/precision.py]: Me too

File: syft/frameworks/torch/nn/__init__.py [hunk 2]
```diff
@@ -1,11 +1,24 @@
...
+from syft.generic.frameworks.overload import overloaded
+
+
+@overloaded.module
+def nn(module):
+    """
+    The syntax is the same, so @overloaded.module handles recursion
+    Note that we don't need to add the @staticmethod decorator
+    """
+    module.Conv2d = Conv2d
```
Comment (Jasopaum) [inline syft/frameworks/torch/nn/__init__.py]: nit: can you put this line next to the one with AvgPool2d?

File: syft/frameworks/torch/nn/functional.py [hunk 3]
```diff
@@ -121,18 +125,24 @@ def conv2d(input, weight, bias=None, stride=1, padding=0, dilation=1, groups=1):
...
             res.append(tmp)
-        res = th.cat(res, dim=2)
+        res = torch.cat(res, dim=2)
     else:
         res = im_reshaped.matmul(weight_reshaped)
 
     # Add a bias if needed
     if bias is not None:
-        res += bias
+        if bias.is_wrapper:
```
Comment (Jasopaum) [inline syft/frameworks/torch/nn/functional.py]: I think you can simplify this `if`s and `else`s

File: syft/frameworks/torch/rnn.py [hunk 4]
```diff
@@ -1,11 +1,17 @@
+# TODO Workaround to cyclic import
+# HELP NEEDED HERE
+# This file is normally inside the nn/ directory. In the precision.py file
+# when we import nn, the nn/__init__.py gets called. Currently all other classes
+# (ie Conv2d, AvgPool2d, etc) behave well when imported in the nn/__init__.py
+# But due to the RNN Classes requiring the import of FixedPrecisionTensor (line 14)
+# and precision.py requiring the import of nn module, there is a problem.
```
Comment (Jasopaum) [inline syft/frameworks/torch/rnn.py]: Do you think we could remove the import of FixedPrecisionTensor (not in this PR, of course)? I don't think RNN should deal with things at that low a level.

**Pull request merged**